File: src/test/java/OsmPlacesSourceTest.java
import com.google.common.collect.Lists;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import uk.co.eelpieconsulting.osm.nominatim.model.Place;
import uk.co.eelpieconsulting.osm.nominatim.psql.OsmDAO;
import uk.co.eelpieconsulting.osm.nominatim.psql.OsmPlacesSource;
import uk.co.eelpieconsulting.osm.nominatim.psql.PlaceRowParser;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class OsmPlacesSourceTest {

    private static final String DATABASE_HOST = "10.0.45.11";  // TODO inject
    private static final String DATABASE_USER = "www-data";
    private static final String DATABASE_PASSWORD = "";

    private OsmDAO osmDAO;
    private PlaceRowParser placeRowParser;

    @BeforeAll
    public void setup() throws SQLException {
        placeRowParser = new PlaceRowParser();
        osmDAO = new OsmDAO(DATABASE_USER, DATABASE_PASSWORD, DATABASE_HOST);
    }

    @Test
    public void somePlacesSpanMultipleRowsWithTypeAndCategories() throws SQLException {
        ResultSet r = osmDAO.getPlace(4599, "R");

        List<String> types = Lists.newArrayList();
        List<String> categories = Lists.newArrayList();
        while (r.next()) {
            Place place = placeRowParser.buildPlaceFromCurrentRow(r);
            types.add(place.getType());
            categories.add(place.getClassification());
        }

        assertEquals(2, types.size());
        assertEquals( Lists.newArrayList("attraction" , "government"), types);
        assertEquals( Lists.newArrayList("tourism" , "office"), categories);
    }

    @Test
    public void canIterateThroughPlaces() throws Exception {
        OsmPlacesSource osmPlacesSource = new OsmPlacesSource(osmDAO, placeRowParser, "R");
        int rowsInterated = 0;
        while (osmPlacesSource.hasNext() && rowsInterated < 1000) {
            osmPlacesSource.next();
            rowsInterated++;
        }

        assertEquals(1000, rowsInterated);
    }

    @Test
    public void canExtractDisplayAddressForPlace() throws Exception {
        final ResultSet placeRow = osmDAO.getPlace(284926920, "W");
        placeRow.next();

        final Place place = placeRowParser.buildPlaceFromCurrentRow(placeRow);

        assertEquals("Twickenham Rowing Club, Church Lane, Cole Park, Twickenham, London Borough of Richmond upon Thames, London, Greater London, England, TW1 3DY, United Kingdom", place.getAddress());
    }

    @Test
    public void placeTagsShouldIncludeTheClassificationTypeAndExtraTags() throws Exception {
        final ResultSet placeRow = osmDAO.getPlace(284926920, "W");
        placeRow.next();

        final Place place = placeRowParser.buildPlaceFromCurrentRow(placeRow);

        assertTrue(place.getTags().contains("leisure|sports_centre"));
        assertTrue(place.getTags().contains("sport|rowing"));
    }

}

File: src/test/kotlin/uk/co/eelpieconsulting/osm/nominatim/psql/OsmPlacesSourceTest.kt
package uk.co.eelpieconsulting.osm.nominatim.psql

import org.junit.Assert.*
import org.junit.Test
import uk.co.eelpieconsulting.osm.nominatim.model.Place

class OsmPlacesSourceTest {

    // Default credentials for Nominatim 3.7 Docker image; nothing to see here
    private val DATABASE_HOST = "localhost:5432"  // TODO inject
    private val DATABASE_USER = "nominatim"
    private val DATABASE_PASSWORD = "qaIACxO6wMR3"

    private val osmDAO = OsmDAO(DATABASE_USER, DATABASE_PASSWORD, DATABASE_HOST)
    private val placeRowParser = PlaceRowParser()

    val limitMuchLargeThanExpectedNumberOfRows= 1000L

    @Test
    fun canRetrieveSingleRowPlace() {
        val twickenham = 21099166L

        fun cursor(start: Long, pageSize: Long) = osmDAO.getPlace(twickenham, "N", pageSize)

        val osmPlacesSource = OsmPlacesSource(placeRowParser, ::cursor, limitMuchLargeThanExpectedNumberOfRows)

        var found = emptyList<Place>()
        while (osmPlacesSource.hasNext()) {
            found = found + osmPlacesSource.next()
        }

        assertEquals(1, found.size)

        val firstRow = found.first()
        assertEquals("Twickenham", firstRow.name)
        assertEquals("Twickenham, London Borough of Richmond upon Thames, London, Greater London, England, TW1 3RZ, United Kingdom", firstRow.address)
        assertEquals("place", firstRow.classification)
        assertEquals("suburb", firstRow.type)
        assertEquals(15, firstRow.adminLevel) // Admin levels increase with depth; countries are highest
        assertEquals(20, firstRow.addressRank) // This is an address rank; higher numbers will present further to the left in an en address
    }

    @Test
    fun somePlacesHaveNoName() {
        val chargingStationWithNoName = 6919655077L

        fun cursor(start: Long, pageSize: Long) = osmDAO.getPlace(chargingStationWithNoName, "N", pageSize)

        val osmPlacesSource = OsmPlacesSource(placeRowParser, ::cursor, limitMuchLargeThanExpectedNumberOfRows)

        var found = emptyList<Place>()
        while (osmPlacesSource.hasNext()) {
            found = found + osmPlacesSource.next()
        }

        assertEquals(1, found.size)

        val place = found.first()
        assertNull("Some places have no name", place.name)
        assertEquals("Places with no name still have an address; and this can end up been quite high ranking which is why it's important to filter them out",
            "St Marks Gate, Fish Island, Bow, London Borough of Tower Hamlets, London, Greater London, England, E9 5HP, United Kingdom",
                place.address)

        assertEquals(15, place.adminLevel)
        assertEquals(30, place.addressRank)
    }

    @Test
    fun countriesHaveLowerAdminLevels() {
        val england = 58447L

        fun cursor(start: Long, pageSize: Long) = osmDAO.getPlace(england, "R", pageSize)

        val osmPlacesSource = OsmPlacesSource(placeRowParser, ::cursor, limitMuchLargeThanExpectedNumberOfRows)

        var found = emptyList<Place>()
        while (osmPlacesSource.hasNext()) {
            found = found + osmPlacesSource.next()
        }

        assertEquals(1, found.size)

        val firstRow = found.first()
        assertEquals("England", firstRow.name)
        assertEquals(4, firstRow.adminLevel)
    }

    @Test
    fun canRetrieveRowsForMultiRowPlaces() {
        fun cursor(start: Long, pageSize: Long) = osmDAO.getPlace(16431, "R", pageSize)    // Southsea castle
        val osmPlacesSource = OsmPlacesSource(placeRowParser, ::cursor, limitMuchLargeThanExpectedNumberOfRows)

        var found = emptyList<Place>()
        while (osmPlacesSource.hasNext()) {
            found = found + osmPlacesSource.next()
        }

        assertEquals(2, found.size)
    }

    @Test
    fun canPaginateBeyondTheFirstPage() {
        fun cursor(start: Long, pageSize: Long) = osmDAO.getPlaces(start, pageSize, "R")

        val osmPlacesSource = OsmPlacesSource(placeRowParser, ::cursor, 10)

        var rowsIterated = 0
        val recordCountKnownToExceedPaginationSize = 20
        while (osmPlacesSource.hasNext() && rowsIterated < recordCountKnownToExceedPaginationSize) {
            osmPlacesSource.next()
            rowsIterated++
        }

        assertEquals(recordCountKnownToExceedPaginationSize, rowsIterated)
    }

}